Add a VariantF datatype and a membership proof.

rohit507 opened the pull request:
I've added a module `Type.Set.VariantF` with two major components. 

1) A way to prove that if `ForAllIn c bst` then `c (Follow ss bst)` is true as long as `ss :: [Side]` points to an actual element of `bst`. 

The typeclass `ForAllIn c bst` is equivalent to `forall f. (Member f bst ~ 'True) => c f` and the function `forMember :: SSide ss -> Dict (c (Follow ss bst))` lets you get the corresponding dict. 

2) A `VariantF (bst :: TypeSet (* -> *)) (a :: *) :: *` datatype that can hold `f a` for some `f` in `bst`. 

This comes with instances of `Functor`, `Foldable`, `Traversable`, and an iffy `Show` instance. (The latter because it's really nice for debugging even if there can't be a valid `Read` instance) 

Here's a quick test:
```Haskell
type VF = VariantF (Insert [] (Insert Maybe Empty)) Int

l :: VF
l = toVariantF @[] [1,2,3,4]

b :: VF
b = toVariantF @Maybe (Just 3)

main :: IO ()
main = do 
  print . fromVariantF @[] $ l
  -- Just [1,2,3,4]
  print . fromVariantF @[] $ fmap (+5) l
  -- Just [6,7,8,9]
  traverse print l
  -- 1
  -- 2
  -- 3
  -- 4
  print l
  -- toVariantF @([SomeType]) $ [1,2,3,4]
  print b
  -- toVariantF @(Maybe SomeType) $ Just 3
```

isovector commented on src/Type/Set/VariantF.hs:
```diff
@@ -0,0 +1,194 @@
+-- anything already in the tree.
+proveFollowInsertF :: forall (ss :: [Side]) (f :: * -> *) (bst :: TypeSet (* -> *)).
+  Follow ss (Insert f bst) :~~: Follow ss bst
+proveFollowInsertF = unsafeCoerce HRefl
```
Is this different from the one in `Type.Set.Variant`? Can we reuse the same one?

rohit507 replied:
Technically yes it's using `:~~:` which is an equality proof for arbitrary kinds instead of `:~:` which seems to only work on the types with kind `*`. 

I'm pretty confident it can be replaced, but I was trying to keep the changes localized so it would be easier for you to review and because I'm not sure how you'd want it to be organized.

isovector commented on src/Type/Set/VariantF.hs:
```diff
@@ -0,0 +1,194 @@
+
+------------------------------------------------------------------------------
+-- | Weaken a 'VariantF' so that it can contain something else.
+weakenF :: forall f bst a. VariantF bst a -> VariantF (Insert f bst) a
```
likewise

rohit507 replied:
See https://github.com/isovector/type-sets/pull/8#issuecomment-529173965 . I think the library could probably stop with just `weaken` and `weakenK` with that approach.

isovector commented on src/Type/Set/VariantF.hs:
```diff
@@ -0,0 +1,194 @@
+
+data SomeType = SomeType
+
+-- | I have no good way to ensure that a is typeable so the output replaces
```
What about `instance (Typeable a, ForAllIn (forall x. Show x => Show (f x)) bst) => Show (VariantF bst)`, where you somehow bind `f` in the `ForAllIn`?

rohit507 replied:
If I were just going to skip the `Show1` instance for `Show` I'd just go with `instance (Typeable a, ForAllIn Typeable bst, Show a, ForAllIn Show1 bst) => Show (VariantF bst a)`

isovector commented:
Very cool, thanks for the PR. I have some questions about the wider maintainability of this library (in an ideal world we don't need new versions of `Variant` for every kind!)

rohit507 commented:
> Very cool, thanks for the PR. I have some questions about the wider maintainability of this library (in an ideal world we don't need new versions of `Variant` for every kind!)

Yeah, I'm not sure how to do that within the current bounds of Haskell. I'll play with it more tomorrow, but it would be nice to have something like 

```haskell
data VariantK (bst :: TypeSet (k -> *)) (a :: k) where
  VariantK :: SSide ss -> Follow ss bst a -> VariantK bst a
```

but that doesn't quite work since the ideal interface would let us instantiate that as `VariantK (bst :: TypeSet (* -> * -> *) (a :: *) (b :: *) :: *` with some sort of type level `curry`/`uncurry`.

Though it doesn't seem too hard to just define 'VariantK' and let some newtypes or type synonyms capture the common cases for `k`. That would mean having only `Variant` and `VariantK` defined in the library and moving from there. 

I'll play with it tomorrow and see what I can do.

rohit507 commented:
Hmm, that wouldn't solve the instance problem, but with `ForAllIn` and `forMember` writing those instances follows a simple pattern that shouldn't be hard for people using the library to follow.

  - Unpack all the `ForAllIn` constraints by matching on the `Dict`s you get from `forMember`. 
  - Write the instance pretty much like normal.

rohit507 commented:
Hmm, It works fine on my box. I'm happily using my fork in a separate project. 

I'm pretty sure the error is because the `.cabal` file is included in the repo and hasn't been updated to account for 'constraints' being added to the deps. I just ended up deleting the `.cabal` file at one point and never actually updated it. 

Anyway, major changes: 

  - Moved `ForAllIn` and associated elements to `Type.Set.Variant` where they make more sense 
  - Added `Eq`, `Ord` and `Show` instances to `Variant` and `VariantF` 
  - Generalized the kind of `VariantF` into `VariantF (bst :: TypeSet (k -> *)) (a :: k) :: *`. It matches the `VariantK` idea from above, but because I didn't get something like `VariantP (bst :: TypeSet (k1 -> k2 -> *) (a :: k1) (b :: k2) :: *` working, I didn't change the name.

rohit507 commented:
Yeah, there's some issue with the cabal file and trying to debug from here takes far too long. 

I've got no more time set aside for randomly exploring this, but if you have concrete changes you'd like me to make I'll be happy to do that. Likewise if you've got some specific idea you want me to test.

rohit507 commented:
Also I'd argue that `VariantF` with it's one type parameter is a sufficient stopping point for the library. It's very useful for defining an `EADT` type and working with complex F-Algebras in general. 

I could maybe see an argument for `VariantP` with its two parameters, because of categories, lenses, and (pro|bi)functors. But the net utility of adding `VariantP` seems a lot lower than the net utility for `VariantF`, and the net utility of anything with more than two type params seems vanishingly small.

isovector commented:
Awesome, thanks!
